tests/Cache/FileSystemCacheItemPoolTest.php

```diff
@@ -0,0 +1,229 @@
+    }
+
+    public function testDeleteItems()
+    {
+        $items = [
+            $this->getNewItem(),
+            $this->getNewItem('NewItem2'),
+            $this->getNewItem('NewItem3')
+        ];
+
+        foreach ($items as $item) {
+            $this->pool->save($item);
+        }
+
+        $itemKeys = array_map(function (CacheItemInterface $item) {
+            return $item->getKey();
+        }, $items);
+
+        $result = $this->pool->deleteItems($itemKeys);
+        $this->assertTrue($result);
+    }
+
+    public function testGetItems()
+    {
+        $items = [
+            $this->getNewItem(),
+            $this->getNewItem('NewItem2'),
+            $this->getNewItem('NewItem3')
+        ];
+
+        foreach ($items as $item) {
+            $this->pool->save($item);
+        }
+
+        $keys = array_map(function (CacheItemInterface $item) {
+            return $item->getKey();
+        }, $items);
+        array_push($keys, 'NonExistant');
+
+        $retrievedItems = $this->pool->getItems($keys);
+
+        foreach ($items as $item) {
+            $this->assertTrue($retrievedItems[$item->getKey()]->isHit());
+        }
+
+        $this->assertFalse($retrievedItems['NonExistant']->isHit());
+    }
+
+    public function testClear()
+    {
+        $item = $this->getNewItem();
+        $this->pool->save($item);
+        $this->assertLessThan(scandir($this->defaultCacheDirectory), 2);
+        $this->pool->clear();
+        // Clear removes all the files, but scandir returns `.` and `..` as files
+        $this->assertEquals(count(scandir($this->defaultCacheDirectory)), 2);
+    }
+
+    public function testSaveDeferredAndCommit()
+    {
+        $item = $this->getNewItem();
+        $this->pool->saveDeferred($item);
+        $this->assertFalse($this->pool->getItem($item->getKey())->isHit());
+
+        $this->pool->commit();
+        $this->assertTrue($this->pool->getItem($item->getKey())->isHit());
+    }
+
+    public function testGetItemWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
+            try {
+                $item = $this->getNewItem($char);
+                $this->pool->getItem($item->getKey());
+                $this->fail('The key ' . $char . ' is passing validation when should not be valid');
+            } catch (InvalidArgumentException $e) {
+                $this->assertTrue($e instanceof InvalidArgumentException);
+            }
+        }
+    }
+
+    public function testGetItemsWithIncorrectKeyShouldThrowAnException()
+    {
+        foreach ($this->invalidChars as $char) {
+            try {
+                $item = $this->getNewItem($char);
+                $this->pool->getItems([$item->getKey()]);
+                $this->fail('The key ' . $char . ' is passing validation when should not be valid');
+            } catch (InvalidArgumentException $e) {
+                $this->assertTrue($e instanceof InvalidArgumentException);
```

Comment:
For all of these, use PHPUnit's built-in `expectException` and `expectExceptionMessage` functions, which you should call at the top of the test, e.g.

```php
    public function testGetItemsWithIncorrectKeyShouldThrowAnException()
    {
        $this->expectException(InvalidArgumentException:class);
        $this->expectExceptionMessage('Whatever your expected exception message is');
       // .. continue with the rest of the test
    }

```

Reply:
I actually used that one but because I made this test with this in mind:
If someone edits the list of characters or the `validKey` method gets updated and one of them "passes", using `expectException` passes even when  one character is valid.

For example, I tried proving that this test would fail by adding the char `a` to the invalid characters and the test was passing when in reality should have failed.

I think this might be fixed with the comment below 👇

Reply:
Yes, you will need to split every exception out into a separate test 👍